### PR TITLE
pulling in jacdac-ts

### PR DESCRIPTION
**Pull request by pelikhan**
Complete rewrite of JACDAC to use JACDAC typescript (https://github.com/jacdac/jacdac-ts ).
- [x] SAMD21
- [x] SAMD51 build
- [x] STM32 build
- [ ] NTF build (#ifdef everything out?)

### CODE REVIEW

File: libs/jacdac-ts/makecodeoptions.ts [hunk 1]
```diff
@@ -1,24 +1,28 @@
...
+            console.add(ConsolePriority.Log, message);
+        }
         getSerialNumber() : Buffer {
             const buf = control.createBuffer(8);
             buf.setNumber(NumberFormat.UInt32LE,0,control.deviceSerialNumber())
             return buf;
         }
     }
+
+    export let options: JDOptions = new MakeCodeOptions();
```
Comment (pelikhan) [inline libs/jacdac-ts/makecodeoptions.ts]: you'll need to update package.json reference as well when ready to build

File: libs/jacdac-ts/makecodejacdac.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        }
+    }
+
+    let jacdacStarted = false;
+    export function start() : void
+    {
+        if (jacdacStarted)
+            return;
+
+        jacdacStarted = true;
```
Comment (pelikhan) [inline libs/jacdac-ts/makecodejacdac.ts]: should try to clean all these "starts"

File: libs/base/configkeys.h [hunk 3]
```diff
@@ -1,6 +1,16 @@
...
+// used by pins.cpp to mask off the pin name from any config
+// lower 16 pins of value are the pin name
+#define CFG_PIN_NAME_MSK 0x0000ffff
+// upper 16 bits of value is any configuration of the pin.
+#define CFG_PIN_CONFIG_MSK 0xffff0000
+
+// begin optional pin configurations
+#define CFG_PIN_CONFIG_ACTIVE_LO 0x10000
+
+
```
Comment (jamesadevine) [inline libs/base/configkeys.h]: @mmoskal

Comment (jamesadevine) [reply libs/base/configkeys.h]: https://github.com/Microsoft/pxt-maker/commit/8f096e995bbd09bdb0f7143c73dcf2eea23f1854

change in use.

Comment (jamesadevine) [reply libs/base/configkeys.h]: @pelikhan

File: libs/core/keyvaluestorage.ts [hunk 4]
```diff
@@ -8,7 +8,17 @@ namespace configStorage {
         if (value == null)
             configStorage.removeItem(key);
         else
-            configStorage.setBuffer(key, control.createBufferFromUTF8(value));
+        {
```
Comment (jamesadevine) [inline libs/core/keyvaluestorage.ts]: @pelikhan

File: libs/jacdac/makecodejacdac.ts [hunk 5]
```diff
@@ -70,6 +70,12 @@ namespace jacdac {
             configStorage.setItem(DEVICE_NAME_STORAGE_KEY, name);
         };
 
+        jacdac.JACDAC.instance.onIdentificationRequest = function () {
+            // DEVICE_ID_JACDAC_CONFIGURATION_SERVICE = 33
```
Comment (jamesadevine) [inline libs/jacdac/makecodejacdac.ts]: Couldn't figure out how to refresh dal.d.ts, so I just hacked them in... @pelikhan

File: libs/jacdac-game/debugger.ts [hunk 6]
```diff
@@ -50,6 +76,34 @@ namespace jacdac.dbg {
...
+            screen.print(label, (this.marginx / 2) + horizontalOffset, this.marginy, 1, this.debugFont);
+            screen.print(value, (this.marginx / 2) + horizontalOffset, this.debugFont.charHeight + (2 * this.marginy), 1, this.debugFont);
+
+            horizontalOffset += size + 1
+
+            return horizontalOffset;
+        }
+
+        paintDiagnosticsBar(){
+            const diag = jacdac.diagnostics();
```
Comment (pelikhan) [inline libs/jacdac-game/debugger.ts]: can "diag" be undefined?

Comment (jamesadevine) [reply libs/jacdac-game/debugger.ts]: Yeah, I was planning to fix it up this am ;)

**Pull request merged**